### PR TITLE
fix Invisible products do not update url_rewrite properly

### LINKED ISSUE #19293: Invisible products do not update url_rewrite properly
### Preconditions
1. Magento 2.4-develop

### Steps to reproduce
1. Have a product in the Magento backoffice with the visibility "Not visible individually". Set the URL key of the product to "product-a"
2. Open the product in de backoffice and change the URL Key to "product-b" and check the "create permanent redirect" checkbox
3. Save the product. 
4. Open the product again, change the visibility of the product to "Catalog, search" and change the URL key to "product-c", check the "create permanent redirect" checkbox.
5. Save the product again.

### Expected result
1. Permanent redirects should be present in the url rewrite database table from product-a to product-b
2. Permanent redirects should be present in the url rewrite database table from product-b to product-c
3. Expected result is that URL rewrite of products with "not individually visible" will also update.

### Actual result (*)
1. In the URL rewrite database table, only a references to product-c are available and permanent redirects from product-b to product-c are available. 
2. product-a is not redirecting to product-b
3. URL rewrites for products which where once visible and now are not visible anymore are not being updated anymore and the old URL key which was present when the product was still visible is being kept in the URL rewrite table.

### PR DESCRIPTION
**Pull request by engcom-Charlie**
### Description (*)

Fixed and covered by test

### Related Pull Requests
https://github.com/magento/inventory/pull/3176
https://github.com/magento-commerce/magento2ee/pull/2600

### Fixed Issues (if relevant)

1. Fixes magento/magento2#19293

### Manual testing scenarios (*)

1.see #19293

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

File: app/code/Magento/CatalogUrlRewrite/Observer/ProcessUrlRewriteOnChangeProductVisibilityObserver.php [hunk 1]
```diff
@@ -15,6 +16,8 @@
 
 /**
  * Consider URL rewrites on change product visibility via mass action
+ *
+ * @deprecated Url rewrites will be created for products with all visibility types.
```
Comment (ihor-sviziev) [inline app/code/Magento/CatalogUrlRewrite/Observer/ProcessUrlRewriteOnChangeProductVisibilityObserver.php]: I think this comment isn't clear that this observer will be never called. Also not clear where url rewrites will be created. I thing we need to add something like this:
```suggestion
 * @deprecated Class is not used anymore. Url rewrites will be created for products with all visibility types.
 * @see \Magento\CatalogUrlRewrite\Observer\ProductProcessUrlRewriteSavingObserver
```

File: app/code/Magento/CatalogUrlRewrite/Model/Products/AdaptUrlRewritesToVisibilityAttribute.php [hunk 2]
```diff
@@ -18,7 +19,9 @@
 use Magento\UrlRewrite\Service\V1\Data\UrlRewrite;
 
 /**
- *  Save/Delete UrlRewrites by Product ID's and visibility
+ * Save/Delete UrlRewrites by Product ID's and visibility
+ *
+ * @deprecated Url rewrites will be created for products with all visibility types.
```
Comment (ihor-sviziev) [inline app/code/Magento/CatalogUrlRewrite/Model/Products/AdaptUrlRewritesToVisibilityAttribute.php]:
```suggestion
 * @deprecated Class isn't used anymore. Url rewrites will be created for products with all visibility types.
 * @see \Magento\CatalogUrlRewrite\Observer\ProductProcessUrlRewriteSavingObserver
```

File: app/code/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserver.php [hunk 3]
```diff
@@ -44,32 +49,29 @@ public function __construct(
     ) {
         $this->productUrlRewriteGenerator = $productUrlRewriteGenerator;
         $this->urlPersist = $urlPersist;
-        $this->productUrlPathGenerator = $productUrlPathGenerator ?: ObjectManager::getInstance()
-            ->get(ProductUrlPathGenerator::class);
+        $this->productUrlPathGenerator = $productUrlPathGenerator
+            ?: ObjectManager::getInstance()->get(ProductUrlPathGenerator::class);
```
Comment (ihor-sviziev) [inline app/code/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserver.php]: backward compatibility policy is not needed for observers, I think we can make this argument not optional

File: dev/tests/integration/testsuite/Magento/CatalogUrlRewrite/Model/ProductUrlRewriteGeneratorTest.php [hunk 4]
```diff
@@ -3,59 +3,96 @@
...
+     *
+     * @param ProductInterface $product
+     * @return array
+     */
+    private function getGeneratedUrls(ProductInterface $product): array
+    {
+        $urls = $this->model->generate($product);
+
+        return array_map(
             function ($url) {
```
Comment (ihor-sviziev) [inline dev/tests/integration/testsuite/Magento/CatalogUrlRewrite/Model/ProductUrlRewriteGeneratorTest.php]: Can we define argument and return types here?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
Also please review failed tests, looks like some additional tests should be adjusted

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@engcom-Charlie tests still failing, please check it

**Review by ihor-sviziev — COMMENTED** [review]
@engcom-Charlie tests still failing, please check it

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@engcom-Charlie tests still failing, please check it

**Comment by ihor-sviziev** [comment]
@engcom-Charlie will you be able to update your PR?

**Comment by ihor-sviziev** [comment]
@engcom-Charlie  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@engcom-Charlie @engcom-Hotel 
My review comments still actual, please update the PR https://github.com/magento/magento2/pull/29069#pullrequestreview-447810292

**Comment by engcom-Charlie** [comment]
@ihor-sviziev can you please look through again? I've done it on by https://github.com/magento/magento2/pull/29069/commits/67a691e773a9b6003149885c2afcf6ee133f0816

**Comment by ihor-sviziev** [comment]
@engcom-Charlie yeah, now I resolved all the issues I found, but still tests fails, please review them

**Comment by engcom-Charlie** [comment]
@ihor-sviziev we will fix failed tests.

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento WebAPI Tests, Integration Tests

**Comment by ihor-sviziev** [comment]
@magento run WebAPI Tests, Integration Tests

**Comment by ihor-sviziev** [comment]
@magento run WebAPI Tests, Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run integration tests

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run WebAPI Tests, Functional Tests

**Comment by engcom-Hotel** [comment]
@magento run WebAPI Tests, Functional Tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@engcom-Charlie @engcom-Hotel any updates?

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Static Tests

**Comment by ihor-sviziev** [comment]
@magento run WebAPI Tests, Functional Tests

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests

**Comment by ihor-sviziev** [comment]
@sidolov @gabrieldagama could you add some info why integration & web api tests are failing? There is no info in the reports

**Comment by sidolov** [comment]
```
...........Error in fixture: "\/var\/www\/html\/app\/code\/Magento\/InventoryConfigurableProductIndexer\/Test\/_files\/product_configurable_multiple.php".
 URL key for specified store already exists.
 #0 /var/www/html/app/code/Magento/UrlRewrite/Model/Storage/AbstractStorage.php(87): Magento\UrlRewrite\Model\Storage\DbStorage->doReplace()
#1 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(58): Magento\UrlRewrite\Model\Storage\AbstractStorage->replace()
#2 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(138): Magento\UrlRewrite\Model\Storage\DbStorage\Interceptor->___callParent()
#3 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(153): Magento\UrlRewrite\Model\Storage\DbStorage\Interceptor->Magento\Framework\Interception\{closure}()
#4 /var/www/html/dev/tests/integration/tmp/sandbox--76ccf56fc1cd540134f59b4069e60caaa4132cb3976271c3398ab8e1c4c45409/generated/code/Magento/UrlRewrite/Model/Storage/DbStorage/Interceptor.php(50): Magento\UrlRewrite\Model\Storage\DbStorage\Interceptor->___callPlugins()
#5 /var/www/html/app/code/Magento/CatalogStaging/Model/Url/Storage.php(56): Magento\UrlRewrite\Model\Storage\DbStorage\Interceptor->replace()
#6 /var/www/html/app/code/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserver.php(72): Magento\CatalogStaging\Model\Url\Storage->replace()
#7 /var/www/html/lib/internal/Magento/Framework/Event/Invoker/InvokerDefault.php(88): Magento\CatalogUrlRewrite\Observer\ProductProcessUrlRewriteSavingObserver->execute()
#8 /var/www/html/lib/internal/Magento/Framework/Event/Invoker/InvokerDefault.php(74): Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod()
#9 /var/www/html/app/code/Magento/Staging/Model/Event/Manager.php(97): Magento\Framework\Event\Invoker\InvokerDefault->dispatch()
#10 /var/www/html/dev/tests/integration/tmp/sandbox--76ccf56fc1cd540134f59b4069e60caaa4132cb3976271c3398ab8e1c4c45409/generated/code/Magento/Staging/Model/Event/Manager/Proxy.php(95): Magento\Staging\Model\Event\Manager->dispatch()
#11 /var/www/html/lib/internal/Magento/Framework/Model/AbstractModel.php(829): Magento\Staging\Model\Event\Manager\Proxy->dispatch()
#12 /var/www/html/app/code/Magento/Catalog/Model/Product.php(973): Magento\Framework\Model\AbstractModel->afterSave()
#13 /var/www/html/dev/tests/integration/tmp/sandbox--76ccf56fc1cd540134f59b4069e60caaa4132cb3976271c3398ab8e1c4c45409/generated/code/Magento/Catalog/Model/Product/Interceptor.php(248): Magento\Catalog\Model\Product->afterSave()
#14 /var/www/html/lib/internal/Magento/Framework/EntityManager/Observer/AfterEntitySave.php(34): Magento\Catalog\Model\Product\Interceptor->afterSave()
#15 /var/www/html/lib/internal/Magento/Framework/Event/Invoker/InvokerDefault.php(88): Magento\Framework\EntityManager\Observer\AfterEntitySave->execute()
#16 /var/www/html/lib/internal/Magento/Framework/Event/Invoker/InvokerDefault.php(74): Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod()
#17 /var/www/html/app/code/Magento/Staging/Model/Event/Manager.php(97): Magento\Framework\Event\Invoker\InvokerDefault->dispatch()
#18 /var/www/html/dev/tests/integration/tmp/sandbox--76ccf56fc1cd540134f59b4069e60caaa4132cb3976271c3398ab8e1c4c45409/generated/code/Magento/Staging/Model/Event/Manager/Proxy.php(95): Magento\Staging\Model\Event\Manager->dispatch()
#19 /var/www/html/lib/internal/Magento/Framework/EntityManager/EventManager.php(51): Magento\Staging\Model\Event\Manager\Proxy->dispatch()
#20 /var/www/html/app/code/Magento/Staging/Model/Operation/Create.php(138): Magento\Framework\EntityManager\EventManager->dispatchEntityEvent()
#21 /var/www/html/lib/internal/Magento/Framework/EntityManager/EntityManager.php(96): Magento\Staging\Model\Operation\Create->execute()
#22 /var/www/html/app/code/Magento/Catalog/Model/ResourceModel/Product.php(764): Magento\Framework\EntityManager\EntityManager->save()
#23 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(58): Magento\Catalog\Model\ResourceModel\Product->save()
#24 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(138): Magento\Catalog\Model\ResourceModel\Product\Interceptor->___callParent()
#25 /var/www/html/app/code/Magento/CatalogSearch/Model/Indexer/Fulltext/Plugin/Product.php(58): Magento\Catalog\Model\ResourceModel\Product\Interceptor->Magento\Framework\Interception\{closure}()
#26 /var/www/html/app/code/Magento/CatalogSearch/Model/Indexer/Fulltext/Plugin/Product.php(28): Magento\CatalogSearch\Model\Indexer\Fulltext\Plugin\Product->addCommitCallback()
#27 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(135): Magento\CatalogSearch\Model\Indexer\Fulltext\Plugin\Product->aroundSave()
#28 /var/www/html/lib/internal/Magento/Framework/App/Cache/FlushCacheByTags.php(69): Magento\Catalog\Model\ResourceModel\Product\Interceptor->Magento\Framework\Interception\{closure}()
#29 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(135): Magento\Framework\App\Cache\FlushCacheByTags->aroundSave()
#30 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(153): Magento\Catalog\Model\ResourceModel\Product\Interceptor->Magento\Framework\Interception\{closure}()
#31 /var/www/html/dev/tests/integration/tmp/sandbox--76ccf56fc1cd540134f59b4069e60caaa4132cb3976271c3398ab8e1c4c45409/generated/code/Magento/Catalog/Model/ResourceModel/Product/Interceptor.php(194): Magento\Catalog\Model\ResourceModel\Product\Interceptor->___callPlugins()
#32 /var/www/html/app/code/Magento/Catalog/Model/ProductRepository.php(823): Magento\Catalog\Model\ResourceModel\Product\Interceptor->save()
#33 /var/www/html/app/code/Magento/Catalog/Model/ProductRepository.php(572): Magento\Catalog\Model\ProductRepository->saveProduct()
#34 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(58): Magento\Catalog\Model\ProductRepository->save()
#35 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(138): Magento\Catalog\Model\ProductRepository\Interceptor->___callParent()
#36 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(153): Magento\Catalog\Model\ProductRepository\Interceptor->Magento\Framework\Interception\{closure}()
#37 /var/www/html/dev/tests/integration/tmp/sandbox--76ccf56fc1cd540134f59b4069e60caaa4132cb3976271c3398ab8e1c4c45409/generated/code/Magento/Catalog/Model/ProductRepository/Interceptor.php(41): Magento\Catalog\Model\ProductRepository\Interceptor->___callPlugins()
#38 /var/www/html/app/code/Magento/InventoryConfigurableProductIndexer/Test/_files/product_configurable_multiple.php(67): Magento\Catalog\Model\ProductRepository\Interceptor->save()
#39 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Annotation/AbstractDataFixture.php(87): require('/var/www/html/a...')
#40 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Annotation/AbstractDataFixture.php(121): Magento\TestFramework\Annotation\AbstractDataFixture->_applyOneFixture()
#41 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Annotation/DataFixture.php(35): Magento\TestFramework\Annotation\AbstractDataFixture->_applyFixtures()
#42 [internal function]: Magento\TestFramework\Annotation\DataFixture->startTestTransactionRequest()
#43 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/EventManager.php(49): call_user_func_array()
#44 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Event/Transaction.php(76): Magento\TestFramework\EventManager->fireEvent()
#45 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Event/Transaction.php(46): Magento\TestFramework\Event\Transaction->_processTransactionRequests()
#46 [internal function]: Magento\TestFramework\Event\Transaction->startTest()
#47 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/EventManager.php(49): call_user_func_array()
#48 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Event/PhpUnit.php(127): Magento\TestFramework\EventManager->fireEvent()
#49 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestResult.php(407): Magento\TestFramework\Event\PhpUnit->startTest()
#50 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestResult.php(624): PHPUnit\Framework\TestResult->startTest()
#51 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestCase.php(772): PHPUnit\Framework\TestResult->run()
#52 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(639): PHPUnit\Framework\TestCase->run()
#53 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(639): PHPUnit\Framework\TestSuite->run()
#54 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(639): PHPUnit\Framework\TestSuite->run()
#55 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(639): PHPUnit\Framework\TestSuite->run()
#56 /var/www/html/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(653): PHPUnit\Framework\TestSuite->run()
#57 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(108): PHPUnit\TextUI\TestRunner->run()
#58 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(68): PHPUnit\TextUI\Command->run()
#59 /var/www/html/vendor/phpunit/phpunit/phpunit(61): PHPUnit\TextUI\Command::main()
#60 {main}
```

**Comment by ihor-sviziev** [comment]
Hi @engcom-Charlie @engcom-Hotel I believe now it will be much easier to understand where the issue is.

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run Integration Tests,WebAPI Tests

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests,WebAPI Tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests,WebAPI Tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @engcom-Charlie @engcom-Hotel 
seems like these changes are causing regression issues as you have to add some specific url key, but it will not be added also in the custom extensions -> websites will become broken.
Do we have any other option in order NOT to add these URL keys?

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests, WebAPI Tests

**Comment by engcom-Hotel** [comment]
I have reviewed the provided solution and found that it causes a critical issue with url_key for invisible products. It is not suitable and too risky to make such changes. PR will be closed.
Please feel free to reopen it back if you disagree or have an alternative solution.
Thank you.

**Pull request closed without merging**